File: ruby/palindrome-products/palindrome_products.rb
class Palindromes
  Palindrome = Struct.new :value, :factors

  def initialize(max_factor: , min_factor: 1)
    @max_factor, @min_factor = max_factor, min_factor
  end

  def generate
    @palindromes_with_products = Hash[pairs.zip(products)].select { |_, product| is_palindrome?(product) }
    @palindromes = @palindromes_with_products.values.uniq
  end

  def largest
    max = @palindromes.max
    Palindrome.new max, products_with_palindrome[max]
  end

  def smallest
    min = @palindromes.min
    Palindrome.new min, products_with_palindrome[min]
  end

  private

  def is_palindrome?(number)
    number.to_s == number.to_s.reverse
  end

  def products_with_palindrome
    products_with_palindrome = Hash.new { |hash, key| hash[key] = [] }
    @palindromes_with_products.each_with_object(products_with_palindrome) do |(product, palidrome), hash|
      hash[palidrome] << product
    end
  end

  def pairs
    (@min_factor..@max_factor).to_a.repeated_combination(2)
  end

  def products
    pairs.map { |pair| pair.first * pair.last }
  end

end

File: python/palindrome-products/palindrome_products.py
def largest_palindrome(max_factor, min_factor=0):
    all_palindromes = _palindromes(max_factor, min_factor)
    if len(all_palindromes) > 0:
        max_value = max(all_palindromes)[0]
        return (max_value, [pair[1] for pair in all_palindromes if pair[0] == max_value])
    else:
        return [None, []]


def smallest_palindrome(max_factor, min_factor=0):
    all_palindromes = _palindromes(max_factor, min_factor)
    if len(all_palindromes) > 0:
        min_value = min(all_palindromes)[0]
        return (min_value, [pair[1] for pair in all_palindromes if pair[0] == min_value])
    else:
        return [None, []]

def _is_palindrome(number):
    return list(str(number)) == list(reversed(str(number)))

def _palindromes(max_factor, min_factor=0):
    if min_factor > max_factor or min_factor < 0:
        raise ValueError("min_factor > maw_factor")

    return sorted([(first*second, (first, second))
                   for first in range(min_factor, max_factor+1)
                  for second in range(first, max_factor+1)
                   if _is_palindrome(first*second)])
